### PR TITLE
add unpack_sequence and unpad_sequence functions

### LINKED ISSUE #66549: add unpack_sequence and unpad_sequence functions
## 🚀 Feature

Add inverse function for `torch.nn.utils.rnn.pack_sequence` that takes `PackedSequence` object as input and returns list of variable length `Tensors`.

## Motivation

Working with sequential data with variable length (e.g. working with text or time series using LSTM) the proposed function is almost always needed but you have to implement it yourself.

## Pitch

n/a

## Alternatives

n/a

## Additional context

Usage example ([link](https://gist.github.com/dayyass/b64cdf552576b64c282c3cf06f39b062)):
```python3
import torch
from torch.nn.utils.rnn import pack_sequence, pad_packed_sequence


def unpack_sequence(packed_sequences):
    """Unpacks PackedSequence into a list of variable length Tensors"""
    
    unpacked_sequences = []
    
    padded_sequences, lengths = pad_packed_sequence(packed_sequences, batch_first=True)
    
    max_length = padded_sequences.shape[1]
    idx = torch.arange(max_length)

    for seq, length in zip(padded_sequences, lengths):
        mask = idx < length
        unpacked_seq = seq[mask]
        unpacked_sequences.append(unpacked_seq)
        
    return unpacked_sequences


a = torch.tensor([1,2,3])
b = torch.tensor([4,5])
c = torch.tensor([6])

sequences = [a, b, c]
packed_sequences = pack_sequence(sequences)
unpacked_sequences = unpack_sequence(packed_sequences)

assert all([torch.allclose(a, b) for a, b in zip(sequences, unpacked_sequences)])
```

cc @albanD @mruberry @jbschlosser @walterddr @zou3519

### PR DESCRIPTION
**Pull request by dayyass**
Fixes #66549

### CODE REVIEW

**Review by jbschlosser — COMMENTED** [review]
Thanks for the contribution! Would you also mind adding a test into `test/test_nn.py` under `TestPackedSequence` that verifies that `unpack_sequence()` is a correct inverse of `pack_sequence()` (as demonstrated in your doc example)?

**Comment by dayyass** [comment]
> Thanks for the contribution! Would you also mind adding a test into `test/test_nn.py` under `TestPackedSequence` that verifies that `unpack_sequence()` is a correct inverse of `pack_sequence()` (as demonstrated in your doc example)?

Of course I would add tests.

**Comment by jbschlosser** [comment]
> @jbschlosser, trying to run test local, but met the following exception: `ImportError: cannot import name 'MappingProxyType' from partially initialized module 'types'`
> 
> Do you know, what might cause this error?

@dayyass Hm, haven't seen that one before. It works for me running this from the root source directory:

```
python test/test_nn.py TestNN.test_unpack_sequence
```

Maybe the problem is environment or installation related?

**Review by jbschlosser — COMMENTED** [review]
Any luck getting the tests running locally? I briefly checked out a couple of the failing checks in CI and recommended some fixes below, hopefully that helps!

**Comment by dayyass** [comment]
> Any luck getting the tests running locally? I briefly checked out a couple of the failing checks in CI and recommended some fixes below, hopefully that helps!

I still unable to run tests locally, but I decomposed them in jupyter notebook and ran - all tests passed!

**Review by jbschlosser — APPROVED** [review]
LGTM! Thanks for the comprehensive testing :) Just one minor comment to simplify it a bit

**Comment by dayyass** [comment]
@albanD, what do you think about this PR?

**Review by albanD — COMMENTED** [review]
Looks good to me. @jbschlosser comment above to clean up the test would be nice.

**Comment by dayyass** [comment]
> Looks good to me. @jbschlosser comment above to clean up the test would be nice.

Done!

File: test/test_nn.py [hunk 1]
```diff
@@ -6742,6 +6742,40 @@ def pad(tensor, length):
...
+        a = torch.tensor([1, 2, 3])
+        b = torch.tensor([4, 5])
+        c = torch.tensor([6])
+        sequences = [a, b, c]
+
+        lengths = torch.as_tensor([v.size(0) for v in sequences])
+        for batch_first in [True, False]:
+            padded_sequences = rnn_utils.pad_sequence(sequences, batch_first=batch_first)
+            unpadded_sequences = rnn_utils.unpad_sequence(padded_sequences, lengths, batch_first=batch_first)
+            _compatibility_test(sequences, unpadded_sequences)
```
[the remark this answers is not in the document]
Comment (dayyass) [reply test/test_nn.py]: Done!

**Comment by albanD** [comment]
Thanks!
Could you rebase on top of master to make sure that all the CI runs fine and this should be ready to land!

**Pull request closed without merging**